aopy/precondition.py

```diff
@@ -1,2 +1,225 @@
+    # filtered_data = lfilter(b,a,data)
+    filtered_data = filtfilt(b, a, data)
+    return filtered_data
+
+def bandpass_multitaper_filter_data(data, fs, NW = None, BW= None, adaptive = False, jackknife = True, sides = 'default'):
+    '''
+
+    Args:
+        data (ndarray):  time series data where time axis is assumed to be on the last axis (n_channels , n_samples)
+        fs (float): sampling rate of the signal
+        NW (float): Normalized half bandwidth of the data tapers in Hz
+        BW (float): sampling bandwidth of the data tapers in Hz
+        adaptive (bool): Use an adaptive weighting routine to combine the PSD estimates of different tapers.
+        jackknife (bool): Use the jackknife method to make an estimate of the PSD variance at each point.
+        sides (str): This determines which sides of the spectrum to return.
+
+    Returns:
+        f (ndarray) : Frequency points vector
+        psd_est (ndarray): estimated power spectral density (PSD)
+        nu (ndarray): if jackknife = True; estimated variance of the log-psd. If Jackknife = False; degrees of freedom in a chi square model of how the estimated psd is distributed wrt true log - PSD
+    '''
+    f, psd_mt, nu = tsa.multi_taper_psd(data, fs, NW, BW,  adaptive, jackknife, sides )
+    return f, psd_mt, nu
+
+def multitaper_lfp_bandpower(f,psd_est, bands, n_channels, no_log):
+    '''
+
+    Args:
+        psd_est (ndarray): power spectral density - output of bandpass_multitaper_filter_data
+        bands (list): lfp bands should be a list of tuples representing ranges e.g., bands = [(0, 10), (10, 20), (130, 140)] for 0-10, 10-20, and 130-140 Hz
+        n_channels: number of channels
+        no_log (bool): boolean to select whether lfp band power should be in log scale or not
+
+    Returns:
+        lfp_power (ndarray): lfp band power for each channel for each band specified ( n_channels * n_features, 1)
+    '''
+
+    lfp_power = np.zeros((n_channels * len(bands), 1))
+    small_epsilon = 0
+    fft_inds = dict()
+
+    for band_idx, band in enumerate(bands):
+            fft_inds[band_idx] = [freq_idx for freq_idx, freq in enumerate(f) if band[0] <= freq < band[1]]
+
+    for idx, band in enumerate(bands):
+        if n_channels == 1:
+            if no_log:
+                lfp_power[idx * n_channels: (idx + 1) * n_channels, 0] = np.mean(psd_est[fft_inds[idx]], axis=0)
+            else:
+                lfp_power[idx * n_channels: (idx + 1) * n_channels, 0] = np.mean(np.log10(psd_est[fft_inds[idx]] + small_epsilon), axis=0)
+        else:
+            if no_log:
+                lfp_power[idx * n_channels: (idx + 1) * n_channels, 0] = np.mean(psd_est[:,fft_inds[idx]], axis=1)
+            else:
+                lfp_power[idx * n_channels: (idx + 1) * n_channels, 0] = np.mean(np.log10(psd_est[:,fft_inds[idx]] + small_epsilon), axis=1)
+
+    return lfp_power
+
+def get_psd(data, fs,l):
```

Comment:
should probably be get_psd_welch() so we can later write a wrapper that has 'method' as a parameter